### PR TITLE
Merge develop-meed 

### PR DESCRIPTION
**Pull request by AzmiTouil**

### CODE REVIEW

File: portlets/src/main/webapp/vue-app/user-connectors-setting/js/UserConnectorService.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+export function getUsersConnectorsSetting(username) {
+  return fetch(`${eXo.env.portal.context}/${eXo.env.portal.rest}/gamification/user/connectors?username=${username}`, {
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/user-connectors-setting/js/UserConnectorService.js]:
```suggestion
  return fetch(`${eXo.env.portal.context}/${eXo.env.portal.rest}/gamification/connectors?expand=${expand}`, {
```

File: portlets/src/main/webapp/vue-app/user-connectors-setting/js/UserConnectorService.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    if (resp?.ok) {
+      return resp.json();
+    } else {
+      throw new Error('Error when getting user connectors');
+    }
+  });
+}
+
+export function connect(connectorName, accessToken) {
+  return fetch(`${eXo.env.portal.context}/${eXo.env.portal.rest}/gamification/user/connectors/${connectorName}/connect`, {
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/user-connectors-setting/js/UserConnectorService.js]:
```suggestion
  return fetch(`${eXo.env.portal.context}/${eXo.env.portal.rest}/gamification/connectors/connect`, {
```

File: portlets/src/main/webapp/vue-app/user-connectors-setting/js/UserConnectorService.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        throw new Error('AccountAlreadyUsed');
+      } else {
+        throw new Error('Error while validating access token');
+      }
+    }
+  });
+}
+
+export function disconnect(connectorName) {
+  return fetch(`${eXo.env.portal.context}/${eXo.env.portal.rest}/gamification/user/connectors/${connectorName}`, {
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/user-connectors-setting/js/UserConnectorService.js]:
```suggestion
  return fetch(`${eXo.env.portal.context}/${eXo.env.portal.rest}/gamification/connectors/disconnect`, {
```

File: services/src/main/java/io/meeds/gamification/model/ConnectorLoginRequest.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import java.io.Serializable;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class ConnectorLoginRequest implements Serializable {
```
Comment (boubaker) [inline services/src/main/java/io/meeds/gamification/model/ConnectorLoginRequest.java]:
```suggestion
public class RemoteConnectorConnectRequest implements Serializable {
```

File: services/src/main/java/io/meeds/gamification/plugin/ConnectorPlugin.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * A plugin that will be used by {@link ConnectorService} to add and remove user
+ * gamification connectors
+ */
+public abstract class ConnectorPlugin extends BaseComponentPlugin {
+
+  /**
+   * Validates a user Token with the Gamification Connector provider
+   *
+   * @param accessToken connector access token
+   * @return the connector identifier {@link String}
```
Comment (boubaker) [inline services/src/main/java/io/meeds/gamification/plugin/ConnectorPlugin.java]:
```suggestion
   * @return the user identifier corresponding to Access Token generated on Remote Connector
```

File: portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorCardList.vue [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+-->
+<template>
+  <v-card flat>
+    <div class="text-header-title py-2 py-sm-5 d-flex align-center">{{ $t('gamification.connectors.label.connectors') }}</div>
+    <v-card-text class="text-font-size pb-2 pt-0 pb-sm-5 px-0 d-flex align-center">{{ $t('gamification.connectors.label.valueContributions') }}</v-card-text>
+    <v-card-text class="pa-0">
+      <v-item-group>
+        <v-container class="pa-0">
+          <div v-if="enabledConnectors && enabledConnectors.length" class="ma-0 d-flex flex-wrap">
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorCardList.vue]: To move to parent IMO
```suggestion
          <div class="ma-0 d-flex flex-wrap">
```

File: portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorDetail.vue [hunk 7]
```diff
@@ -0,0 +1,91 @@
+<!--
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorDetail.vue]: Rename files to not have `GamificationAdminConnector` prefix all time. In fact this prefix is useful in component names only (in `initComponents.js`).

File: portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorDetail.vue [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+Lesser General Public License for more details.
+You should have received a copy of the GNU Lesser General Public License
+along with this program; if not, write to the Free Software Foundation,
+Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+-->
+<template>
+  <v-app>
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorDetail.vue]: `v-app` to use in Main component only, see https://v2.vuetifyjs.com/en/components/application/

> v-app should only be rendered within your application ONCE.

```suggestion
```

File: portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorSettings.vue [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  <v-app>
+    <v-card class="my-3 border-radius" flat>
+      <extension-registry-component
+        v-if="displayConnectorSettings && connector"
+        :params="params"
+        :component="connectorExtension" />
+      <gamification-admin-connector-detail
+        v-else-if="displayConnectorDetails && connector"
+        class="px-4"
+        :connector="connector" />
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorSettings.vue]:
```suggestion
      <template v-if="selectedConnector">
        <extension-registry-component
          v-if="editSettings"
          :params="selectedConnector"
          :component="connectorExtension" />
        <gamification-admin-connector-detail
          v-else
          class="px-4"
          :connector="connector" />
      </template>
```

File: portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorSettings.vue [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  computed: {
+    params() {
+      return {
+        apiKey: this.connector?.apiKey,
+        secretKey: this.connector?.secretKey,
+        redirectUrl: this.connector?.redirectUrl || '',
+        enabled: this.connector?.enabled,
+      };
+    },
+  },
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorSettings.vue]:
```suggestion
```

File: portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorSettings.vue [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            const connectorObj = connectorsSettings.find(connectorSettings => connectorSettings.name === connector.componentOptions.name);
+            connector.componentOptions.apiKey = connectorObj?.apiKey || '';
+            connector.componentOptions.secretKey = connectorObj?.secretKey || '';
+            connector.componentOptions.redirectUrl = connectorObj?.redirectUrl || '';
+            connector.componentOptions.enabled = connectorObj?.enabled;
+          });
+        } else {
+          this.adminConnectorsExtensionsSettings.forEach(connector => (connector.enabled = false));
+        }
+        this.connectors = this.adminConnectorsExtensionsSettings.map(connectorsSettingsExtension => connectorsSettingsExtension.componentOptions);
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/admin-connectors-settings/components/GamificationAdminConnectorSettings.vue]:
```suggestion
        this.connectors = this.adminConnectorsExtensionsSettings.map(extension => extension.componentOptions);
```

File: portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  }),
+  computed: {
+    Title() {
+      return this.isCurrentUserProfile ? this.$t('gamification.connectors.profile.YourGamifiedProfiles') : this.$t('gamification.connectors.profile.MyGamifiedProfiles');
+    },
+    enabledConnectors() {
+      return  this.connectors?.filter(connector => connector.enabled) || [];
+    },
+    enabledConnectedConnectors() {
+      return  this.connectors?.filter(connector => connector.enabled && connector.identifier !== '') || [];
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue]:
```suggestion
      return  this.connectors?.filter(connector => connector.enabled && connector.identifier?.length) || [];
```

File: portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    notYetConnected() {
+      return this.enabledConnectors.length > 0 && this.connectors.every(connector => connector.enabled && connector.identifier === '') && this.isCurrentUserProfile;
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue]:
```suggestion
      return this.enabledConnectors.length && !this.enabledConnectedConnectors?.length;
```

File: portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    displayed() {
+      return this.notYetConnected || this.enabledConnectedConnectors.length > 0;
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue]:
```suggestion
      return this.enabledConnectors.length;
```

File: portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  },
+  created() {
+    this.refreshUserConnectorList();
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue]:
```suggestion
    document.addEventListener(`extension-gamification-connectors-updated`, this.refreshUserConnectorList);
    this.refreshUserConnectorList();
```

File: portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    class="white">
+    <v-toolbar
+      id="gamifiedProfilesHeader"
+      color="white"
+      height="64"
+      flat
+      class="border-box-sizing">
+      <div class="d-flex flex py-3">
+        <div class="text-header-title text-sub-title text-no-wrap d-flex align-center">
+          {{ Title }}
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/gamifiedProfiles/components/GamifiedProfiles.vue]:
```suggestion
          {{ title }}
```

File: portlets/src/main/webapp/WEB-INF/conf/gamification/portal/portal-configuration.xml [hunk 17]
```diff
@@ -80,6 +80,119 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
+                    </value-param>
+                    <object-param>
+                        <name>group.configuration</name>
+                        <description>description</description>
+                        <object type="org.exoplatform.portal.config.NewPortalConfig">
+                            <field name="ownerType">
+                                <string>group</string>
+                            </field>
+                            <field name="override">
+                                <boolean>${exo.kudos.portalConfig.metadata.override:true}</boolean>
```
Comment (boubaker) [inline portlets/src/main/webapp/WEB-INF/conf/gamification/portal/portal-configuration.xml]:
```suggestion
                                <boolean>${exo.gamification.portalConfig.metadata.override:true}</boolean>
```
+
XML format please

File: portlets/src/main/webapp/WEB-INF/conf/gamification/portal/portal-configuration.xml [hunk 18]
```diff
@@ -80,6 +80,119 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
+                                <string>war:/conf/gamification/portal</string>
+                            </field>
+                        </object>
+                    </object-param>
+                </init-params>
+            </component-plugin>
     </external-component-plugins>
 
+    <external-component-plugins>
+      <target-component>org.exoplatform.commons.addons.AddOnService</target-component>
```
Comment (boubaker) [inline portlets/src/main/webapp/WEB-INF/conf/gamification/portal/portal-configuration.xml]: Please move this configuration to `dynamic-container-configuration.xml`

File: services/src/main/resources/db/changelog/gamification.db.changelog-1.0.0.xml [hunk 19]
```diff
@@ -697,4 +697,20 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
+        <column name="CONNECTOR_NAME" type="NVARCHAR(250)">
+          <constraints nullable="false"/>
+        </column>
+        <column name="REMOTE_ID" type="NVARCHAR(250)"/>
+        <column name="USER_ID" type="BIGINT">
+          <constraints nullable="false"/>
+        </column>
+      </createTable>
+      <addUniqueConstraint columnNames="CONNECTOR_NAME, REMOTE_ID" tableName="GAMIFICATION_CONNECTOR_ACCOUNTS"/>
+      <createSequence sequenceName="SEQ_GAMIFICATION_CONNECTOR_ACCOUNTS_ID" startValue="1" />
```
Comment (boubaker) [inline services/src/main/resources/db/changelog/gamification.db.changelog-1.0.0.xml:716]:
```suggestion
      <createSequence sequenceName="SEQ_GAMIFICATION_CONNECTOR_ACCOUNTS_ID" startValue="1" />
```

Please add this into a dedicated changeSet in order to apply it on some Rdbms only, as it's not useful for MySQL by example.

File: services/src/main/java/io/meeds/gamification/service/impl/ConnectorSettingServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+    String encryptedApiKey;
+    String encryptedSecretKey;
+    try {
+      encryptedApiKey = codecInitializer.getCodec().encode(remoteConnectorSettings.getApiKey());
+      encryptedSecretKey = codecInitializer.getCodec().encode(remoteConnectorSettings.getSecretKey());
+      remoteConnectorSettings.setApiKey(encryptedApiKey);
+      remoteConnectorSettings.setSecretKey(encryptedSecretKey);
+    } catch (TokenServiceInitializationException e) {
+      LOG.warn("Error generating connector settings", e);
+    }
```
Comment (boubaker) [inline services/src/main/java/io/meeds/gamification/service/impl/ConnectorSettingServiceImpl.java]:
```suggestion
    try {     remoteConnectorSettings.setSecretKey(codecInitializer.getCodec().encode(remoteConnectorSettings.getSecretKey()));
    } catch (TokenServiceInitializationException e) {
      throw new IllegalStateException("Error encrypting Secret Key", e);
    }
    remoteConnectorSettings.setApiKey(getApiKey);
```

Comment (AzmiTouil) [reply services/src/main/java/io/meeds/gamification/service/impl/ConnectorSettingServiceImpl.java]: Done!

**Review by boubaker — APPROVED** [review]

**Pull request merged**